### PR TITLE
Fastapi

### PR DESCRIPTION
**Pull request by ernstdevreede**

### CODE REVIEW

File: python/python-fastapi-server/routers/healthcheck.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+"""handleRouteHealthCheck"""
+from .setup_adaguc import setup_adaguc
+from fastapi import Request, APIRouter
+
+healthCheckRouter = APIRouter(responses={404: {"description": "Not found"}})
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/healthcheck.py]: Is that not the default response for404?

File: python/python-fastapi-server/routers/healthcheck.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+
+@healthCheckRouter.get("/healthcheck")
+async def handleRouteHealthCheck(request: Request):
+    """handleRouteHealthCheck"""
+    setup_adaguc(False)
+    response = {"response": "OK"}
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/healthcheck.py]: Use Response object (like in other places)?

File: python/python-fastapi-server/routers/wmswcs.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+"""wmsWcsRouter"""
+from .setup_adaguc import setup_adaguc
+from fastapi import Request, APIRouter, Response
+
+from .setup_adaguc import setup_adaguc
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/wmswcs.py]: Duplicate import

File: python/python-fastapi-server/routers/wmswcs.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import sys
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/wmswcs.py]: Not used.

File: python/python-fastapi-server/routers/wmswcs.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import os
+import logging
+import time
+
+wmsWcsRouter = APIRouter(responses={404: {"description": "Not found"}})
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/wmswcs.py]: Is that not the default response?

File: python/python-fastapi-server/routers/wmswcs.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+@wmsWcsRouter.get("/wms")
+@wmsWcsRouter.get("/wcs")
+@wmsWcsRouter.get("/adagucserver")
+@wmsWcsRouter.get("/adaguc-server")
+async def handleWMS(
+    req: Request,
+    response: Response,
+):
+    start = time.perf_counter()
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/wmswcs.py]: start and stage3 time are not used.

File: python/python-fastapi-server/routers/wmswcs.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    adagucInstance = setup_adaguc()
+    logger.info("instance:" + str(adagucInstance))
+    url = req.url
+
+    logger.info(req.url)
+    stage1 = time.perf_counter()
+
+    adagucenv = {}
+
+    """ Set required environment variables """
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/wmswcs.py]: Are these really intended to be docstrings and not comments?

File: python/python-fastapi-server/routers/wmswcs.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    responseCode = 200
+    if status != 0:
+        logger.info("Adaguc status code was %d" % status)
+        responseCode = 500
+    response = Response(content=data.getvalue(), status_code=responseCode)
+
+    stage3 = time.perf_counter()
+
+    # Append the headers from adaguc-server to the headers from flask.
```
Comment (lukas-phaf) [inline python/python-fastapi-server/routers/wmswcs.py]:
```suggestion
    # Append the headers from adaguc-server to the headers from fastapi.
```

File: python/python-fastapi-server/main.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import os
+
+from fastapi import Depends, FastAPI, Request
```
Comment (lukas-phaf) [inline python/python-fastapi-server/main.py]: Depends not used.

File: python/python-fastapi-server/main.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    return response
+
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
```
Comment (lukas-phaf) [inline python/python-fastapi-server/main.py]: Is this what we had in Flask? 
I don't think origins=* together with allows_credentials is a good idea.
Actually, according to the documentation this is not possible:
`allow_credentials - Indicate that cookies should be supported for cross-origin requests. Defaults to False. Also, allow_origins cannot be set to ['*'] for credentials to be allowed, origins must be specified.`

Do you get an error from this?

File: python/python-fastapi-server/main.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+if "EXTERNALADDRESS" in os.environ:
+    app.add_middleware(FixSchemeMiddleware)
```
Comment (lukas-phaf) [inline python/python-fastapi-server/main.py]: Why is this hack needed?

File: python/python-fastapi-server/main.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@app.get("/")
+async def root():
+    return {"message": "ADAGUC server base URL, use /wms, /wcs, /autowms or /ogcapi"}
```
Comment (lukas-phaf) [inline python/python-fastapi-server/main.py]: Mention 'adagucopendap' here?

**Comment by lukas-phaf** [comment]
Old flask code is still there, remove it?

File: doc/Developing.md [hunk 13]
```diff
@@ -75,10 +75,10 @@ export ADAGUC_DB="user=adaguc password=adaguc host=localhost dbname=adaguc"
 export ADAGUC_ENABLELOGBUFFER=FALSE
 # To enable core dump generation, additionally do:
 #ulimit -c unlimited
-#sudo sysctl -w kernel.core_pattern=core-adagucserver # 
+#sudo sysctl -w kernel.core_pattern=core-adagucserver #
 # Then you can use gdb ./bin/adagucserver core-adagucserver
 
-python3 ./python/python-adaguc-server/main.py
+python3 ./python/python_fastapi_server/main.py
```
Comment (maartenplieger) [inline doc/Developing.md:92]: Can you also change at line 60

pip3 install flask flask-cors flask-caching gunicorn pytest marshmallow owslib pyproj==2.6.1 apispec apispec-webframeworks marshmallow-oneofschema defusedxml werkzeug netCDF4 pillow lxml chardet

to use with the requirements file?


`pip3 install -r requirements.txt`

File: python/python-fastapi-server/main.py (hunk 11 above)
Comment (ernstdevreede) [reply python/python-fastapi-server/main.py]: This is a way to make sure that returned URLs have the correct protocol part (http/https) as this could be lost afgter proxying.

File: python/python-fastapi-server/main.py (hunk 10 above)
Comment (ernstdevreede) [reply python/python-fastapi-server/main.py]: No error seen.

File: Dockerfile [hunk 14]
```diff
@@ -75,10 +75,11 @@ COPY --from=0 /adaguc/adaguc-server-master/python /adaguc/adaguc-server-master/p
 COPY --from=0 /adaguc/adaguc-server-master/tests /adaguc/adaguc-server-master/tests
 COPY --from=0 /adaguc/adaguc-server-master/runtests.sh /adaguc/adaguc-server-master/runtests.sh
 COPY --from=0 /adaguc/adaguc-server-master/requirements.txt /adaguc/adaguc-server-master/requirements.txt
+COPY --from=0 /adaguc/adaguc-server-master/requirements-dev.txt /adaguc/adaguc-server-master/requirements-dev.txt
 
 # Upgrade pip and install python requirements.txt
 RUN pip3 install --no-cache-dir --upgrade pip \
-   && pip3 install --no-cache-dir -r requirements.txt
+   && pip3 install --no-cache-dir -r requirements.txt -r requirements-dev.txt
```
Comment (lukas-phaf) [inline Dockerfile]: Do we want this in our production image?

File: requirements-dev.txt [hunk 15]
```diff
@@ -0,0 +1,4 @@
+pytest
+black
+isort
+pylint
```
Comment (lukas-phaf) [inline requirements-dev.txt:4]: Are these run automatically or by hand?

File: python/python_fastapi_server/routers/middleware.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+class FixSchemeMiddleware:
+    """Fix https scheme behind proxy"""
+
+    def __init__(self, app: ASGIApp):
+        self.app = app
+        self.scheme = None
+        external_address = os.environ["EXTERNALADDRESS"]
+        if external_address:
+            self.scheme = external_address[: external_address.index(":")]
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/middleware.py]: Maybe safer to use this?
https://docs.python.org/3/library/urllib.parse.html

File: python/python_fastapi_server/routers/ogcapi.py [hunk 17]
```diff
@@ -0,0 +1,676 @@
...
+            rel="alternate",
+            title="This document in HTML",
+            type="text/html",
+        )
+    )
+    links.append(
+        Link(
+            href=str(req.url_for("get_conformance")),
+            rel="conformance",
+            title="This document in HTML",
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py]: Title incorrect.

File: python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+    )
+
+
+class Collections(BaseModel):
+    links: List[Link]
+    collections: List[Collection]
+
+
+class LandingPage(BaseModel):
+    title: Optional[str] = Field(None, example="Buildings in Bonn")
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py]: Are you okay with keeping the examples as "buildings"?

File: python/python_fastapi_server/routers/ogcapi.py [hunk 19]
```diff
@@ -0,0 +1,676 @@
...
+                "request": req,
+                "title": "conformance",
+                "description": "D",
+                "conformance": jsonable_encoder(conf_classes),
+            },
+        )
+    return conf_classes
+
+
+def make_open_api():
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py:307]: What trick happens in this function?

File: python/python_fastapi_server/routers/ogcapi.py [hunk 20]
```diff
@@ -0,0 +1,676 @@
...
+    coord = list(map(float, point.split(",")))
+    dimspec = ""
+    if len(dims):
+        for dim in dims.split("|"):
+            dimname, dimval = dim.split("=")
+            dimspec += f"&{dimname}={dimval}"
+
+    datetime_ = datetime_.replace("$", "/")
+    request_url = (
+        f"http://localhost:8000/wms?dataset={collection}&query_layers={observed_property_name}"
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py]: localhost:8000 because it points to the adaguc python wrapper running from the same Python program?

File: python/python_fastapi_server/routers/ogcapi.py [hunk 21]
```diff
@@ -0,0 +1,676 @@
...
+    base_url: str,
+    point=None,
+    bbox=None,
+    datetime_=None,
+    result_time=None,
+    observed_property_name=None,
+    dims=None,
+    npoints=None,
+) -> List[FeatureGeoJSON]:
+    features: List[FeatureGeoJSON] = []
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py:410]: features is redefined later, this one is never used.

File: python/python_fastapi_server/routers/ogcapi.py [hunk 22]
```diff
@@ -0,0 +1,676 @@
...
+    else:
+        coords = [point]
+    if not observed_property_name:
+        collinfo = get_parameters(coll)
+        observed_property_name = [collinfo["layers"][0]["name"]]
+        # Default observedPropertyName = first layername
+    param_list = ",".join(observed_property_name)
+
+    if result_time:
+        collinfo = get_parameters(coll)
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py]: Never used?

File: python/python_fastapi_server/routers/ogcapi.py [hunk 23]
```diff
@@ -0,0 +1,676 @@
...
+        features = []
+        for data in response_data:
+            data_features = feature_from_dat(data, collection, url, True)
+            features.extend(data_features)
+        return features[0]
+
+    return None
+
+
+def get_features_for_items(
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py:399]: Can this (partially) reuse `get_single_item()` function?

File: python/python_fastapi_server/routers/ogcapi.py [hunk 24]
```diff
@@ -0,0 +1,676 @@
...
+    response_model_exclude_none=True,
+)
+async def get_items_for_collection(
+    coll: str,
+    req: Request,
+    response: Response,
+    wanted_format: str = "json",
+    limit: Union[int, None] = Query(default=10),
+    start: Union[int, None] = Query(default=0),
+    bbox: Union[str, None] = Depends(check_bbox),
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py]: Does the type of bbox make sense here, as the Depends(check_bbox) return a list of floats?

File: python/python_fastapi_server/routers/ogcapi.py [hunk 25]
```diff
@@ -0,0 +1,676 @@
...
+        dimname, dimval = dim.split(":")
+        if dimname.upper() == "ELEVATION":
+            dimensions["ELEVATION"] = dimval
+        else:
+            dimensions[dimname] = dimval
+
+    return dimensions
+
+
+def check_bbox_crs(
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py]: Function not used?

File: python/python_fastapi_server/routers/ogcapi.py [hunk 26]
```diff
@@ -0,0 +1,676 @@
...
+
+    if bbox_crs not in SUPPORTED_CRS_LIST:
+        # Error
+        raise HTTPException(
+            status_code=400, detail=f"bbox-crs {bbox_crs} not in supported list"
+        )
+    return bbox_crs
+
+
+def check_crs(crs: Union[str, None] = Query(default=None)) -> str:
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi.py:534]: Function not used?

File: python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+class Extent(BaseModel):
+    spatial: Optional[Spatial] = Field(
+        None, description="The spatial extent of the features in the collection."
+    )
+    temporal: Optional[Temporal] = Field(
+        None, description="The temporal extent of the features in the collection."
+    )
+
+
+class Type(Enum):
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py:79]: Beautiful type names.

File: python/python_fastapi_server/routers/ogcapi_tools.py [hunk 28]
```diff
@@ -0,0 +1,446 @@
...
+    return None
+
+
+def get_datasets(adaguc_data_set_dir):
+    """
+    Return all possible OGCAPI feature datasets, based on the dataset directory
+    """
+    dataset_files = [
+        f
+        for f in os.listdir(adaguc_data_set_dir)
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi_tools.py]: There is now way to ask adaguc about this? Or the adaguc database?

File: python/python_fastapi_server/routers/ogcapi_tools.py [hunk 29]
```diff
@@ -0,0 +1,446 @@
...
+    stage2 = time.perf_counter()
+    logger.info("[PERF] Adaguc execution took: %f", (stage2 - stage1))
+
+    if len(logfile) > 0:
+        logger.info(logfile)
+
+    return status, data
+
+
+# @cacher.memoize(timeout=30)
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/ogcapi_tools.py:124]: This used the flash cache library?
Looks like this is a mature standalone memoization library:
https://pypi.org/project/cachetools/

File: python/python_fastapi_server/routers/wmswcs.py [hunk 30]
```diff
@@ -0,0 +1,98 @@
...
+logger = logging.getLogger(__name__)
+
+
+@wmsWcsRouter.get("/wms")
+@wmsWcsRouter.get("/wcs")
+@wmsWcsRouter.get("/adagucserver")
+@wmsWcsRouter.get("/adaguc-server")
+async def handle_wms(
+    req: Request,
+    response: Response,
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/wmswcs.py]: `response` is not used. Why is it here?

File: python/python_fastapi_server/routers/opendap.py [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+
+from .setup_adaguc import setup_adaguc
+
+opendapRouter = APIRouter(responses={404: {"description": "Not found"}})
+
+logger = logging.getLogger(__name__)
+
+
+@opendapRouter.get("/adagucopendap/{opendappath:path}")
+def handle_opendap(req: Request, opendappath: str):
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/opendap.py:15]: This does pretty much the same as the /adagucserver endpoint. Why does it exist?

File: python/lib/adaguc/AdagucTestTools.py [hunk 32]
```diff
@@ -147,6 +168,17 @@ def compareGetCapabilitiesXML(self, testresultFileLocation, expectedOutputFileLo
...
+        try:
+            for child in obj1.findall("Service/ContactInformation")[0]:
+                if child.text.strip() == "":
+                    child.getparent().remove(child)
+            for child in obj2.findall("Service/ContactInformation")[0]:
+                if child.text.strip() == "":
+                    child.getparent().remove(child)
+        except Exception:
+            pass
+
```
Comment (lukas-phaf) [inline python/lib/adaguc/AdagucTestTools.py:261]: Does this behave differently in your FastAPI implementation?

File: python/python_fastapi_server/routers/templates/ogcapi/item.html [hunk 33]
```diff
@@ -17,15 +17,15 @@ <h3>Feature</h3>
         {% for ts in feature.properties.timestep %}
           {{ ts }} {{feature.properties.result[loop.index0]}} {{feature.geometry.coordinates}} [{{feature.id}}]<br/>
         {% endfor %}
-        <br/>At: {{items.timeStamp}}
+        <br/>At: {{item.timeStamp}}
```
Comment (lukas-phaf) [inline python/python_fastapi_server/routers/templates/ogcapi/item.html:20]: Why the changes in html templates?

File: tests/expectedoutputs/TestWCS/test_WCSGetCapabilities_testdatanc [hunk 34]
```diff
@@ -22,36 +22,36 @@
     <GetCapabilities>
       <DCPType>
         <HTTP>
-          <Get><OnlineResource xlink:type="simple" xlink:href="%26source%3Dtestdata%2Enc&amp;SERVICE=WCS&amp;" /></Get>
+          <Get><OnlineResource xlink:type="simple" xlink:href="source=testdata%2Enc&amp;SERVICE=WCS&amp;" /></Get>
```
Comment (lukas-phaf) [inline tests/expectedoutputs/TestWCS/test_WCSGetCapabilities_testdatanc:25]: All the test result changes seem to be dropping the leading `&`. Is that intended?

File: python/python_fastapi_server/routers/ogcapi_tools.py [hunk 35]
```diff
@@ -0,0 +1,446 @@
...
+
+
+def get_dimensions(layer, skip_dims=None):
+    """
+    get_dimensions
+    """
+    dims = []
+    for dim_name in layer.dimensions:
+        if not dim_name in skip_dims:
+            dim_name = {
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/ogcapi_tools.py]: Please do not use dim_name again, make a new variable for this object.

File: python/python_fastapi_server/routers/ogcapi_tools.py [hunk 36]
```diff
@@ -0,0 +1,446 @@
...
+# @cacher.cached(timeout=30, key_prefix="collections")
+def generate_collections():
+    """
+    Generate OGC API Feature collections
+    """
+    collections = get_datasets(os.environ.get("ADAGUC_DATASET_DIR"))
+    return collections
+
+
+def get_dimensions(layer, skip_dims=None):
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/ogcapi_tools.py]: Please describe the skip_dims option.

File: python/python_fastapi_server/routers/ogcapi_tools.py [hunk 37]
```diff
@@ -0,0 +1,446 @@
...
+
+# @cacher.memoize(timeout=30)
+def get_parameters(collname):
+    """
+    get_parameters
+    """
+    contents = get_capabilities(collname)
+    layers = []
+    for layer in contents:
+        dims = get_dimensions(contents[layer], ["time"])
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/ogcapi_tools.py:182]: This gets all dimensions except for the time dimension?

File: Dockerfile (hunk 14 above)
Comment (maartenplieger) [reply Dockerfile]: Needed for pytest, but is removed afterwards.

File: requirements-dev.txt (hunk 15 above)
Comment (maartenplieger) [reply requirements-dev.txt:4]: These are run by the IDE

File: python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py (hunk 18 above)
Comment (maartenplieger) [reply python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py]: Removed.

File: python/python_fastapi_server/routers/ogcapi.py (hunk 19 above)
Comment (maartenplieger) [reply python/python_fastapi_server/routers/ogcapi.py:307]: The openapi generation does not completely work as it should. This adjusts the openapi object

File: python/python_fastapi_server/routers/ogcapi.py (hunk 20 above)
Comment (maartenplieger) [reply python/python_fastapi_server/routers/ogcapi.py]: Changed to 8080

File: python/python_fastapi_server/routers/middleware.py (hunk 16 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/middleware.py]: Done

File: python/python_fastapi_server/routers/ogcapi.py (hunk 21 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi.py:410]: Fixed

File: python/python_fastapi_server/routers/ogcapi.py (hunk 22 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi.py]: Removed

File: python/python_fastapi_server/routers/ogcapi.py (hunk 24 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi.py]: Not sure.
From FastAPI docs it looked like it has to work this way: the union[str, None] defines an optional string parameter, the Depends checks if it is a string of 4 floats (and returns that).

File: python/python_fastapi_server/routers/ogcapi.py (hunk 26 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi.py:534]: Nope, but it will be used in the future.

File: python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py (hunk 27 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/models/ogcapifeatures_1_model.py:79]: That's what you get with these automatic conversion tools...

File: python/python_fastapi_server/routers/ogcapi.py (hunk 25 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi.py]: Nope, but it will be used in the future.

File: python/python_fastapi_server/routers/ogcapi_tools.py (hunk 28 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi_tools.py]: Nope, adaguc_data_set_dir is the place where the dataset XML files live, should always be available.
This info is not in the database.

File: python/python_fastapi_server/routers/ogcapi_tools.py (hunk 29 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi_tools.py:124]: Using cachetools now.

File: python/python_fastapi_server/routers/wmswcs.py (hunk 30 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/wmswcs.py]: Removed

File: python/python_fastapi_server/routers/opendap.py (hunk 31 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/opendap.py:15]: I guess so it can be switched off and on separately?
Historically grown this way is probably the best explanation.
Let's leave it like this.

File: python/python_fastapi_server/routers/templates/ogcapi/item.html (hunk 33 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/templates/ogcapi/item.html:20]: items was a typo, the parameter is called item

File: python/lib/adaguc/AdagucTestTools.py (hunk 32 above)
Comment (ernstdevreede) [reply python/lib/adaguc/AdagucTestTools.py:261]: I guess so.

File: tests/expectedoutputs/TestWCS/test_WCSGetCapabilities_testdatanc (hunk 34 above)
Comment (ernstdevreede) [reply tests/expectedoutputs/TestWCS/test_WCSGetCapabilities_testdatanc:25]: Yes. Has to do with the fix for over-eagerly urlencoding.

File: python/python_fastapi_server/routers/ogcapi_tools.py (hunk 35 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi_tools.py]: Opps. Sloppy.
Fixed.

File: python/python_fastapi_server/routers/ogcapi_tools.py (hunk 37 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi_tools.py:182]: Yes

File: python/python_fastapi_server/routers/ogcapi_tools.py (hunk 36 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi_tools.py]: Done

File: python/python_fastapi_server/routers/ogcapi.py (hunk 24 above)
Comment (maartenplieger) [reply python/python_fastapi_server/routers/ogcapi.py]: Can we resolve this? Does it work as intended?

**Review by maartenplieger — DISMISSED** [review]
OK

File: python/python_fastapi_server/routers/ogcapi.py (hunk 24 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/ogcapi.py]: I was able to fix this after all.

**Review by maartenplieger — APPROVED** [review]
